### PR TITLE
Capture collection errors (e.g. import failures) in JSON report

### LINKED ISSUE #106: Collection errors (e.g. import failures) are silently dropped from the JSON report
## Summary

When a test file fails to import (or has any other collection-phase error), the Buildkite test collector produces a JSON report with **no entry** for the failed file. Combined with pytest-xdist and bktec's retry mechanism, this causes builds to report **green** even though a test file failed to collect.

## Root cause

The plugin only implements `pytest_runtest_*` hooks, which fire during the test **execution** phase. When a test file has an import error, pytest fails during **collection** — the `pytest_collectreport` hook fires, but the plugin doesn't implement it. The test never enters the run phase, so no `pytest_runtest_*` hooks fire, and the failure is absent from the JSON report.

Three things interact to make this silent in CI:

1. **test-collector-python** has no `pytest_collectreport` hook — the collection error is missing from the JSON report
2. **pytest-xdist** (`-n N`) automatically continues past collection errors and changes the exit code from 2 to 1 — no `--continue-on-collection-errors` flag needed
3. **bktec** rescues exit code 1, parses the JSON (which has no failures), retries any flaky tests, and exits 0

## Reproduction

Full end-to-end reproduction with three scenarios (vanilla pytest, xdist, bktec + retry):

https://github.com/altana-ai/bktec-collection-error-repro

## Expected behavior

Collection errors should appear in the JSON report as failed test entries, so that bktec sees the failure, the error appears in the Buildkite Tests tab, and the build fails.

## Proposed fix

Implement `pytest_collectreport` on `BuildkitePlugin` to capture collection errors (where `report.failed` is true) and add them to the payload as failed `TestData` entries.

### PR DESCRIPTION
**Pull request by jasonwbarnett**
### Changes

Adds a `pytest_collectreport` hook to `BuildkitePlugin` so that collection-phase errors (import failures, syntax errors, etc.) are captured as failed `TestData` entries in the JSON report.

Previously, the plugin only implemented `pytest_runtest_*` hooks, which fire during the test **execution** phase. When a test file failed to import, pytest failed during **collection** and the error was silently dropped from the report.

Combined with pytest-xdist (which continues past collection errors and changes the exit code from 2 to 1) and bktec's retry mechanism, this caused builds to report green even when a test file failed to collect.

### What's included

- `pytest_collectreport` method on `BuildkitePlugin` (~40 lines)
- 5 unit tests for the new hook (passed/failed/empty-nodeid/already-started/tuple-longrepr)
- 3 integration tests (subprocess pytest with a broken import file)
- Sample test file with a deliberate broken import for integration tests

### Known side-effect with bktec retries

When bktec retries failed tests, it constructs the retry node ID as `scope::name` ([pytest.go line 106](https://github.com/buildkite/test-engine-client/blob/v2.4.0/internal/runner/pytest.go#L106)). For normal test failures this produces a valid pytest node ID like `tests/foo.py::test_bar`.

However, for collection errors the scope is empty — and correctly so. A collection error nodeid is just the file path (e.g. `tests/foo.py`) with no `::` separators. When the collector splits on `::`, it gets `scope=""` and `name="tests/foo.py"`. There is no scope because the test never got collected into a class or module hierarchy.

The problem is that bktec unconditionally joins them:

```go
Path: fmt.Sprintf("%s::%s", test.Scope, test.Name)
// "" + "::" + "tests/foo.py" = "::tests/foo.py" — invalid node ID
```

This causes `no tests ran` on the retry attempt. The build still correctly fails — the collection error is reported as a failure and bktec exits non-zero. The retry just can't resolve it (which is expected — an import error can't be fixed by re-running).

**Suggested bktec fix** — handle empty scope when constructing the retry path:

```go
if test.Scope != "" {
    Path: fmt.Sprintf("%s::%s", test.Scope, test.Name)
} else {
    Path: test.Name
}
```

This is a bktec-side issue and outside the scope of this PR.

### Reproduction

Full end-to-end reproduction (pytest, xdist, bktec): https://github.com/altana-ai/bktec-collection-error-repro

Fixes #106

### CODE REVIEW

**Comment by jasonwbarnett** [comment]
@nprizal would appreciate a review here 🙏

**Comment by jasonwbarnett** [comment]
### Self-review notes — known limitations

A few items identified during review that don't warrant code changes but are worth documenting:

**`location` field lacks line number** — For regular tests, `location` is `file.py:42`. For collection errors it's just `file.py` since there's no meaningful test-level line number (the error line is in `failure_expanded` via the traceback).

**Near-zero duration** — Collection errors are created and immediately finished, so `duration` is a few microseconds. This is technically correct (no test execution occurred) but may look odd in analytics dashboards.

**Unit tests use string/tuple longrepr** — Real collection errors produce `CollectErrorRepr` (a `TerminalRepr` subclass), which takes the `_handle_default_longrepr` → `str()` → `_handle_string_longrepr` path in `failure_reasons()`. The unit tests pass plain strings and tuples which exercise other branches. The integration tests do cover the real `CollectErrorRepr` path end-to-end.

**No xdist integration test** — The dedup guard was added to handle xdist workers each firing `pytest_collectreport` for the same file, but there is no integration test running pytest with `-n`. The unit test for dedup verifies the logic directly, and the xdist interaction is well-understood from [the reproduction repo](https://github.com/altana-ai/bktec-collection-error-repro).

**Comment by jasonwbarnett** [comment]
### Design question: should collection errors go into the test results?

I wanted to flag something I'm not sure about and would love maintainer input on.

This PR puts collection errors into `payload.data` — the same bucket as real test results. That data flows to both the JSON file (which bktec reads locally) and the Buildkite API (which feeds Test Engine analytics, the Tests tab, and billing).

I'm wondering if this could have unintended side effects — for example, creating phantom test entries like `tests/foo.py` in Test Engine that count toward billing, never have a passing run, and pollute analytics with things that aren't really tests.

I'm honestly not sure what the right approach is here. A few options I can think of:

1. **This PR's approach** — treat collection errors as failed test results. Simple, and bktec immediately sees the failure. But it does mean non-test entries end up in Test Engine.

2. **Fix in bktec instead** — bktec already sees pytest's exit code (1). The issue is really that bktec's retry logic overwrites the original error when retried tests pass. Maybe bktec could detect "pytest exited non-zero but my JSON shows no failures" and propagate the error. That would keep the collector clean, though it's a change in a different repo.

3. **Separate section in the JSON** — write collection errors to a distinct key (e.g. `"collection_errors"`) so bktec can see them without them flowing into analytics. Would need a coordinated change in both repos.

4. **Tag and filter** — keep collection errors in the results but mark them distinctly so `api.submit()` can skip them.

I don't have enough context on how Test Engine handles these entries or what the billing implications are, so I'd really appreciate your thoughts on which direction makes the most sense. Happy to rework the PR however you'd like.

**Comment by jasonwbarnett** [comment]
Filed a companion issue for the bktec retry node ID bug mentioned in the PR description: https://github.com/buildkite/test-engine-client/issues/485

That tracks the empty-scope case where bktec constructs `::tests/foo.py` instead of `tests/foo.py` when retrying collection errors.

**Comment by niceking** [comment]
Hey @jasonwbarnett thanks for raising this fix! I verified that it worked against the reproduction repo, thanks! Here's the amp thread for internal Buildkite readers: https://ampcode.com/threads/T-019daea4-79e5-77cb-98e5-02d311c1a8d5

> Design question: should collection errors go into the test results?

Yeah great question! I think yes, collection errors should go into test results. Expanding the remit of what test engine tracks is on the cards so collection errors definitely aligns with that. There are some billing implications as you pointed out, but the blast radius of this isn't huge (I can't imagine the magnitude of collection errors being *that* huge), and we bill on the 90th percentile of active tests in a month so I don't think it would have a huge impact. 

I do think that collection errors should be tagged though - both from a metrics and filtering viewpoint if they are ingested into test engine, or so customer could choose not to ingest them. Could you please add a default tag for collection results as part of this PR?

> companion issue for the bktec retry node ID bug

If a test fails on collection... is there any point in retrying this? Is an empty scope a sure fire way of identifying collection errors, or is it possible to have a failing executed test that also has an empty scope? My thinking is perhaps if the empty scope signals collection errors, we could remove these files from the retry list completely.

**Comment by jasonwbarnett** [comment]
Thanks for the quick review and guidance @niceking!

> Could you please add a default tag for collection results as part of this PR?

Done — collection errors are now tagged with `test.collection_error: "true"`. Added assertion coverage in both the unit and integration tests.

> If a test fails on collection... is there any point in retrying this?

No, retrying a collection error is pointless — an import failure won't resolve by re-running. The test file needs a code fix.

> Is an empty scope a sure fire way of identifying collection errors, or is it possible to have a failing executed test that also has an empty scope?

Good question. An empty scope means the nodeid had no `::` separator — just a bare file path like `tests/foo.py`. For executed tests, pytest always produces nodeids with `::` (e.g. `tests/foo.py::test_func` or `tests/foo.py::TestClass::test_method`), so scope is never empty for those.

That said, using the `test.collection_error` tag is probably a more robust signal than relying on empty scope — it's explicit and won't break if some edge case produces an unusual nodeid format. bktec could check for that tag when deciding whether to retry.

File: src/buildkite_test_collector/pytest_plugin/buildkite_plugin.py [hunk 1]
```diff
@@ -48,6 +49,58 @@ def pytest_collection_modifyitems(self, config, items):
...
+
+        failure_reason, failure_expanded = failure_reasons(longrepr=report.longrepr)
+        logger.debug('-> collection error: %s', failure_reason)
+
+        test_data = test_data.failed(
+            failure_reason=failure_reason,
+            failure_expanded=failure_expanded,
+        )
+        test_data = test_data.tag_execution(
+            "test.collection_error", "true"
```
Comment (jasonwbarnett) [inline src/buildkite_test_collector/pytest_plugin/buildkite_plugin.py:98]: @niceking Let me know if I misinterpreted your request for a default tag and/or if this key is rubbish.

Comment (niceking) [inline src/buildkite_test_collector/pytest_plugin/buildkite_plugin.py]: Ah, could we change the name of the tag to include something that signals it's python specific? 

```suggestion
            "test.pytest_collection_error", "true"
```

**Review by niceking — COMMENTED** [review]
Just the one comment about the tag name! 

> That said, using the test.collection_error tag is probably a more robust signal than relying on empty scope 

Yeah let's do that as the bktec side fix - not retrying a test file if there's a collection error. Could you update the bktec to do that too 🙏

**Review by niceking — APPROVED** [review]
